multiprocess_functions.py

```diff
@@ -0,0 +1,166 @@
+
+  sorted_size_dict = dict(sorted(size_dict.items(), key=lambda item: item[1], reverse=True))
+
+  total_cores = get_total_cores()
+  cores_per_process = total_cores // max_processes
+  pool = mp.Pool(processes=max_processes)
+
+  queue = [i for i in range(max_processes)]
+  core_list, numa_nodes = get_core_list(cores_per_process)
+  error_files = []
+  def update_queue(result):
+    print(result)
+    index = core_list.index(result[3][0])
+    queue.append(index // cores_per_process)
+    # queue.append(result[3][0] // cores_per_process)
+    if (result[0] != 0):
+      error_files.append(result[1])
+
+  # Iterate over the files and start a new subprocess for each file.
+  print(len(sorted_size_dict))
+  results = [None] * len(sorted_size_dict)
+
+  i = 0
+  for file, value in sorted_size_dict.items():
+    file_path = file
+    process_num = queue.pop(0)
+
+    if max_processes < numa_nodes:
+      if max_processes == 1:
+        if numa_nodes > 1:
+          mem = '0-{}'.format(numa_nodes-1)
+        else:
+          mem = '0'
+      else:
+        mem = '{}-{}'.format(str(process_num * (numa_nodes//max_processes)), str(((process_num + 1) * (numa_nodes//max_processes)) - 1))
+    else:
+      mem = str(process_num//(max_processes//numa_nodes))
+    
+    results[i] = pool.apply_async(bash_subprocess, args=(file_path, mem, core_list[process_num*cores_per_process: (process_num+1)*cores_per_process]), callback = update_queue)
+    i += 1
+    while len(queue) == 0 and i < len(sorted_size_dict):
+        time.sleep(0.05)
+  pool.close()
+  pool.join()
+
+  return error_files
+
+def create_process_list(files, MIN_MEM_PER_PROCESS, MIN_CORES_PER_PROCESS, LOAD_BALANCE_FACTOR):
+  total_cores = get_total_cores()
+  #numa_nodes
+  lscpu = subprocess.Popen(["lscpu"], stdout=subprocess.PIPE)
+  grep = subprocess.Popen(["grep", "NUMA node(s):"], stdin=lscpu.stdout, stdout=subprocess.PIPE)
+  awk = subprocess.Popen(["awk", "{print $3}"], stdin=grep.stdout, stdout=subprocess.PIPE)
+  #Get the output
+  numa_nodes = int(awk.communicate()[0])
+  cores_per_numa = total_cores//numa_nodes
+
+  #sockets
+  lscpu = subprocess.Popen(["lscpu"], stdout=subprocess.PIPE)
+  grep = subprocess.Popen(["grep", "Socket(s):"], stdin=lscpu.stdout, stdout=subprocess.PIPE)
+  awk = subprocess.Popen(["awk", "{print $2}"], stdin=grep.stdout, stdout=subprocess.PIPE)
+  #Get the output
+  sockets = int(awk.communicate()[0])
+
+  memory_per_numa_domain = check_available_memory()/numa_nodes
+  # Memory max
+  mm = memory_per_numa_domain // MIN_MEM_PER_PROCESS
+  # Core_max
+  cm = cores_per_numa // MIN_CORES_PER_PROCESS
+
+  pn = len(files) // numa_nodes
+  # Load_balance_max
+  lbm = max(pn//LOAD_BALANCE_FACTOR, 1)
+  print("Memory max {}, Core max {}, Load Balance max {}".format(mm, cm, lbm))
+  max_p = min(mm, cm, lbm)
+
+  # number which is 2^x and less than or equal to max
+  max_p = 2**int(math.log2(max_p))
+
+  max_processes_list = []
```

Comment:
I think we can improve this logic a bit. Let us leave a TODO here to improve this logic. I think it makes sense to divide by 2 when the numbers are large. But once we are down to, say 4 per NUMA domain, it may be better to do 4,3,2,1.

Reply:
To keep track of this TODO, we can also create an issue saying we should try 4,3,2,1.

Reply:
done